test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs

```diff
@@ -65,122 +48,260 @@ public class InitializeTestMachine : IDisposable
                         }
-                        IISConfigUtility.IsIISReady = true;
                     }
+
+                    _globalTestFlags = _globalTestFlags.ToLower();
                 }
-                catch (Exception ex)
+
+                return _globalTestFlags;
+            }
+        }
+
+        public void InitializeIISServer()
+        {
+            // Check if IIS server is installed or not
+            bool isIISInstalled = true;
+            if (!File.Exists(Path.Combine(IISConfigUtility.Strings.IIS64BitPath, "iiscore.dll")))
+            {
+                isIISInstalled = false;
+            }
+
+            if (!File.Exists(Path.Combine(IISConfigUtility.Strings.IIS64BitPath, "config", "applicationhost.config")))
+            {
+                isIISInstalled = false;
+            }
+            
+            if (!isIISInstalled)
+            {
+                throw new System.ApplicationException("IIS server is not installed");
+            }
+
+            // Check websocket is installed
+            if (File.Exists(Path.Combine(IISConfigUtility.Strings.IIS64BitPath, "iiswsock.dll")))
+            {
+                TestUtility.LogInformation("Websocket is installed");
+            }
+            else
+            {
+                throw new System.ApplicationException("websocket module is not installed");
+            }
+
+            // Clean up IIS worker process
+            TestUtility.ResetHelper(ResetHelperMode.KillWorkerProcess);
+
+            // Reset applicationhost.config
+            TestUtility.LogInformation("Restoring applicationhost.config");
+            IISConfigUtility.RestoreAppHostConfig(restoreFromMasterBackupFile: true);
+            TestUtility.StartW3svc();
+
+            // check w3svc is running after resetting applicationhost.config
+            if (IISConfigUtility.GetServiceStatus("w3svc") == "Running")
+            {
+                TestUtility.LogInformation("W3SVC service is restarted after restoring applicationhost.config");
+            }
+            else
+            {
+                throw new System.ApplicationException("WWW service can't start");
+            }
+
+            // check URLRewrite module exists
+            if (File.Exists(Path.Combine(IISConfigUtility.Strings.IIS64BitPath, "rewrite.dll")))
+            {
+                TestUtility.LogInformation("Verified URL Rewrite module installed for IIS server");
+            }
+            else
+            {
+                throw new System.ApplicationException("URL Rewrite module is not installed");
+            }
+
+            if (IISConfigUtility.ApppHostTemporaryBackupFileExtention == null)
+            {
+                throw new System.ApplicationException("Failed to backup applicationhost.config");
+            }
+        }
+
+        public InitializeTestMachine()
+        {
+            _referenceCount++;
```

Comment:
This is a weird way of doing the singleton pattern. Maybe just have a bool instead or make a new object?

Reply:
I will revisit this and remove this. Eventually this will be gone because the singleton logic won't be necessary when the code refactoring is done.